core/tee/tee_sql_fs.c

```diff
@@ -491,14 +428,161 @@ static TEE_Result sql_fs_write(struct tee_file_handle *fh, size_t pos,
 	if (pos > meta->length)
 		meta->length = pos;
 
-exit:
+	return TEE_SUCCESS;
+}
+static TEE_Result sql_fs_write(struct tee_file_handle *fh, size_t pos,
+			       const void *buf, size_t len)
+{
+	TEE_Result res;
+	struct sql_fs_fd *fdp = (struct sql_fs_fd *)fh;
+
+	mutex_lock(&sql_fs_mutex);
+	sql_fs_begin_transaction_rpc();
+
+	res = sql_fs_write_primitive(fh, pos, buf, len);
+
 	if (res == TEE_SUCCESS)
 		res = tee_fs_htree_sync_to_storage(&fdp->ht);
+
 	sql_fs_end_transaction_rpc(res != TEE_SUCCESS);
 	mutex_unlock(&sql_fs_mutex);
 	return res;
 }
 
+static void sql_fs_close(struct tee_file_handle **fh)
+{
+	struct sql_fs_fd *fdp = (struct sql_fs_fd *)*fh;
+
+	if (fdp) {
+		tee_fs_htree_close(&fdp->ht);
+		tee_fs_rpc_close(OPTEE_MSG_RPC_CMD_SQL_FS, fdp->fd);
+		free(fdp);
+		*fh = NULL;
+	}
+}
+
+static TEE_Result open_internal(struct tee_pobj *po, bool create,
+				struct tee_file_handle **fh)
+{
+	TEE_Result res;
+	struct sql_fs_fd *fdp;
+
+	fdp = calloc(1, sizeof(*fdp));
+	if (!fdp)
+		return TEE_ERROR_OUT_OF_MEMORY;
+	fdp->fd = -1;
+
+	if (create)
+		res = tee_fs_rpc_create(OPTEE_MSG_RPC_CMD_SQL_FS, po, &fdp->fd);
+	else
+		res = tee_fs_rpc_open(OPTEE_MSG_RPC_CMD_SQL_FS, po, &fdp->fd);
+	if (res != TEE_SUCCESS)
+		goto out;
+
+	res = tee_fs_htree_open(create, &sql_fs_storage_ops, fdp, &fdp->ht);
+out:
+	if (res == TEE_SUCCESS) {
+		*fh = (struct tee_file_handle *)fdp;
+	} else {
+		if (fdp && fdp->fd != -1)
+			tee_fs_rpc_close(OPTEE_MSG_RPC_CMD_SQL_FS, fdp->fd);
+		if (create)
+			tee_fs_rpc_remove(OPTEE_MSG_RPC_CMD_SQL_FS, po);
+		free(fdp);
+	}
+
+	return res;
+}
+
+static TEE_Result sql_fs_open(struct tee_pobj *po, struct tee_file_handle **fh)
+{
+	TEE_Result res;
+
+	mutex_lock(&sql_fs_mutex);
+	res = open_internal(po, false, fh);
+	mutex_unlock(&sql_fs_mutex);
+
+	return res;
+}
+
+static TEE_Result sql_fs_create(struct tee_pobj *po, bool overwrite,
+				const void *head, size_t head_size,
+				const void *attr, size_t attr_size,
+				const void *data, size_t data_size,
+				struct tee_file_handle **fh)
+{
+	struct sql_fs_fd *fdp;
+	TEE_Result res;
+	size_t pos = 0;
+	bool old_temporary;
+
+	mutex_lock(&sql_fs_mutex);
+	sql_fs_begin_transaction_rpc();
+	old_temporary = po->temporary;
+
+	res = open_internal(po, true, fh);
+	if (res)
+		goto out;
+
+	if (head && head_size) {
+		res = sql_fs_write_primitive(*fh, pos, head, head_size);
+		if (res)
+			goto out;
```

Comment:
Thanks, closing is enough here as `sql_fs_end_transaction_rpc(res != TEE_SUCCESS);` will roll everything back on failure.